controllers/controllers/clusters/vsphere.go

```diff
@@ -135,5 +137,49 @@ func (v *VSphereClusterReconciler) Reconcile(ctx context.Context, cluster *anywh
 		return reconciler.Result{}, err
 	}
 
+	workerNodeGroupMachineSpecs := make(map[string]anywherev1.VSphereMachineConfigSpec, len(machineConfigMap))
+	for _, wnConfig := range cluster.Spec.WorkerNodeGroupConfigurations {
+		workerNodeGroupMachineSpecs[wnConfig.MachineGroupRef.Name] = machineConfigMap[wnConfig.MachineGroupRef.Name].Spec
+	}
+
+	cp := machineConfigMap[specWithBundles.Spec.ControlPlaneConfiguration.MachineGroupRef.Name]
+	var etcd *anywherev1.VSphereMachineConfigSpec
+	if specWithBundles.Spec.ExternalEtcdConfiguration != nil {
+		etcdSpec := machineConfigMap[specWithBundles.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name]
+		etcd = &etcdSpec.Spec
+	}
+
+	templateBuilder := vsphere.NewVsphereTemplateBuilder(&dataCenterConfig.Spec, &cp.Spec, etcd, workerNodeGroupMachineSpecs, time.Now, true)
+	clusterName := cluster.ObjectMeta.Name
+
+	cpOpt := func(values map[string]interface{}) {
+		values["controlPlaneTemplateName"] = templateBuilder.CPMachineTemplateName(clusterName)
+		controlPlaneUser := machineConfigMap[cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name].Spec.Users[0]
+		values["vsphereControlPlaneSshAuthorizedKey"] = controlPlaneUser.SshAuthorizedKeys[0]
+		etcdUser := machineConfigMap[cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name].Spec.Users[0]
```

Comment:
what if `ExternalEtcdConfiguration` does not exist? (stacked etcd)

Reply:
added a check